ddl/multi_schema_change.go

```diff
@@ -56,18 +57,135 @@ func (d *ddl) MultiSchemaChange(ctx sessionctx.Context, ti ast.Ident) error {
 	return d.callHookOnChanged(job, err)
 }
 
-func checkMultiSchemaInfo(info *model.MultiSchemaInfo, t table.Table) error {
-	err := checkOperateSameColAndIdx(info)
-	if err != nil {
-		return err
+func onMultiSchemaChange(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	if job.MultiSchemaInfo.Revertible {
+		// Handle the rolling back job.
+		if job.IsRollingback() {
+			// Rollback/cancel the sub-jobs in reverse order.
+			for i := len(job.MultiSchemaInfo.SubJobs) - 1; i >= 0; i-- {
+				sub := job.MultiSchemaInfo.SubJobs[i]
+				if sub.IsFinished() {
+					continue
+				}
+				proxyJob := sub.ToProxyJob(job)
+				ver, err = w.runDDLJob(d, t, proxyJob)
+				sub.FromProxyJob(proxyJob)
+				if i == 0 && sub.IsFinished() {
+					job.State = model.JobStateRollbackDone
+				}
+				return ver, err
+			}
+			// The last rollback/cancelling sub-job is done.
+			job.State = model.JobStateRollbackDone
+			return ver, nil
+		}
+
+		// The sub-jobs are normally running.
+		// Run the first executable sub-job.
+		for _, sub := range job.MultiSchemaInfo.SubJobs {
+			if !sub.Revertible || sub.IsFinished() {
+				// Skip the sub-jobs which related schema states
+				// are in the last revertible point.
+				// If a sub job is finished here, it should be a noop job.
+				continue
+			}
+			proxyJob := sub.ToProxyJob(job)
+			ver, err = w.runDDLJob(d, t, proxyJob)
+			sub.FromProxyJob(proxyJob)
+			handleRevertibleException(job, sub, proxyJob.Error)
+			return ver, err
+		}
+
+		// Save table info and sub-jobs for rolling back.
+		tblInfo, _ := t.GetTable(job.SchemaID, job.TableID)
+		subJobs := make([]model.SubJob, len(job.MultiSchemaInfo.SubJobs))
+		// Step the sub-jobs to the non-revertible states all at once.
+		for i, sub := range job.MultiSchemaInfo.SubJobs {
+			if sub.IsFinished() {
+				continue
+			}
+			subJobs[i] = *sub
+			proxyJob := sub.ToProxyJob(job)
+			ver, err = w.runDDLJob(d, t, proxyJob)
+			sub.FromProxyJob(proxyJob)
+			if err != nil || proxyJob.Error != nil {
+				for j := i - 1; j >= 0; j-- {
+					job.MultiSchemaInfo.SubJobs[j] = &subJobs[j]
+				}
+				handleRevertibleException(job, sub, proxyJob.Error)
+				// The TableInfo and sub-jobs should be restored
+				// because some schema changes update the transaction aggressively.
+				return updateVersionAndTableInfo(d, t, job, tblInfo, true)
```

Comment:
This function is executed on every state of every `subJob`. In this case, `genSchemSchemSpanning` is executed one more time for each `subJob` state. Why do we need to call it here?

Reply:
Or can we just do 'txn.Reset' here?